src/main/java/net/mcreator/ui/minecraft/states/JStateLabel.java

```diff
@@ -0,0 +1,156 @@
+	private final Supplier<Stream<JStateLabel>> otherStates;
+
+	private StateMap stateMap = new StateMap();
+	private boolean allowEmpty = false;
+	private NumberMatchType numberMatchType = NumberMatchType.EQUAL;
+
+	private final JTextField label = new JTextField();
+	private final TechnicalButton edit = new TechnicalButton(UIRES.get("16px.edit.gif"));
+
+	public JStateLabel(MCreator mcreator, Supplier<List<PropertyData<?>>> properties,
+			Supplier<Stream<JStateLabel>> otherStates) {
+		super(new BorderLayout(5, 0));
+		this.mcreator = mcreator;
+		this.properties = properties;
+		this.otherStates = otherStates;
+
+		setBackground((Color) UIManager.get("MCreatorLAF.BLACK_ACCENT"));
+		setBorder(BorderFactory.createEmptyBorder(5, 5, 5, 5));
+
+		label.setEditable(false);
+		label.setOpaque(false);
+		refreshState();
+
+		JScrollPane statePane = new JScrollPane(label);
+		statePane.setOpaque(false);
+		statePane.getViewport().setOpaque(false);
+		statePane.setPreferredSize(new Dimension(300, 30));
+		add("Center", statePane);
+
+		edit.setOpaque(false);
+		edit.setMargin(new Insets(0, 0, 0, 0));
+		edit.setBorder(BorderFactory.createEmptyBorder());
+		edit.setContentAreaFilled(false);
+		edit.setToolTipText(L10N.t("components.state_label.edit"));
+		edit.addActionListener(e -> editState());
+
+		JPanel controls = new JPanel();
+		controls.setBackground((Color) UIManager.get("MCreatorLAF.LIGHT_ACCENT"));
+		controls.add(edit);
+		add("East", PanelUtils.centerInPanelPadding(controls, 2, 2));
+	}
+
+	@Override public void setEnabled(boolean enabled) {
+		super.setEnabled(enabled);
+		edit.setEnabled(enabled);
+	}
+
+	public boolean editState() {
+		List<PropertyData<?>> propertyList = properties.get();
+		if (propertyList == null)
+			return false;
+
+		StateMap stateMap = StateEditorDialog.open(mcreator, propertyList, getStateMap(), numberMatchType);
+		if (stateMap == null)
+			return false;
+
+		if (stateMap.isEmpty() && !allowEmpty) {
+			JOptionPane.showMessageDialog(mcreator, L10N.t("components.state_label.error_empty"),
+					L10N.t("components.state_label.error_empty.title"), JOptionPane.ERROR_MESSAGE);
+			return false;
+		} else if (otherStates.get().anyMatch(s -> s != this && s.getStateMap().equals(stateMap))) {
+			JOptionPane.showMessageDialog(mcreator, L10N.t("components.state_label.error_duplicate"),
+					L10N.t("components.state_label.error_duplicate.title"), JOptionPane.ERROR_MESSAGE);
+			return false;
+		}
+
+		setStateMap(stateMap);
+		return true;
+	}
+
+	public JStateLabel setAllowEmpty(boolean allowEmpty) {
+		this.allowEmpty = allowEmpty;
+		return this;
+	}
+
+	public JStateLabel setNumberMatchType(NumberMatchType numberMatchType) {
+		this.numberMatchType = Objects.requireNonNullElse(numberMatchType, NumberMatchType.EQUAL);
+		refreshState();
+		return this;
+	}
+
+	public StateMap getStateMap() {
+		return stateMap;
+	}
+
+	public void setStateMap(StateMap stateMap) {
+		this.stateMap = new StateMap();

```

Comment:
Any particular reason stateMap is not just re-assigned?

Reply:
We use the object directly from the mod element, so it seemed to me changing a state would change it in the mod element as well, and even if the changes are discarded, the state would appear modified the next time when opening it. Tested it now, doesn't happen.

Reply:
✔️ Fixed